### PR TITLE
Introduce DataLoader

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
@l-bat @apertovs @RodionovF @KSTconsult Have a look, please

**Comment by TolyaTalamanov** [comment]
I believe it will a bit clearer how to use it when sample is finished.

File: src/dataloader.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+deepworks::DataLoader::DataLoader(deepworks::IDataset::Ptr dataset,
+                                  int                      batch_size,
+                                  bool                     shuffle)
+    : m_dataset(dataset),
+      m_bs(batch_size),
+      m_shuffle(shuffle) {
+}
+
+bool deepworks::DataLoader::pull(deepworks::Tensor& X, deepworks::Tensor& y) {
+    auto Xshape = makeShape(m_bs, m_dataset->shape().first);
```
Comment (apertovs) [inline src/dataloader.cpp:43]: X_shape
y_shape

File: src/io/image_reader.cpp [hunk 2]
```diff
@@ -138,13 +135,10 @@ void ReadPngFile(std::string_view path, deepworks::Tensor& out_tensor) {
         if (out_tensor.empty()) {
             out_tensor.allocate(deepworks::Shape{height, width, channels});
         } else {
-            DeepWorks_Assert(height == out_tensor.shape()[0]);
-            DeepWorks_Assert(width == out_tensor.shape()[1]);
-            DeepWorks_Assert(channels == out_tensor.shape()[2]);
+            DeepWorks_Assert(height * width * channels == out_tensor.total());
```
Comment (apertovs) [inline src/io/image_reader.cpp:138]: Why should we delete it? These are not equivalent checks

File: include/deepworks/dataloader.hpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+class DataLoader {
+public:
+    DataLoader(IDataset::Ptr dataset, int batch_size = 1, bool shuffle = true);
+    bool pull(Tensor& X, Tensor& y);
+    void reset();
+
+private:
+    IDataset::Ptr m_dataset;
+    int           m_bs;
```
Comment (apertovs) [inline include/deepworks/dataloader.hpp:27]: m_batchsize

File: src/dataloader.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    std::uniform_int_distribution<> dist(0, upper_bound);
+    while (unique.size() != num) {
+        unique.insert(dist(gen));
+    }
+
+    return {unique.begin(), unique.end()};
+}
+
+static std::vector<int> seqIndices(int num, int start) {
```
Comment (apertovs) [inline src/dataloader.cpp:28]: inline?

Comment (TolyaTalamanov) [reply src/dataloader.cpp:28]: what the purpose using `inline` ?

File: src/io/image_reader.cpp (hunk 2 above)
Comment (TolyaTalamanov) [reply src/io/image_reader.cpp:138]: Not make reshape all time, just read image to memory

File: include/deepworks/dataloader.hpp (hunk 3 above)
Comment (TolyaTalamanov) [reply include/deepworks/dataloader.hpp]: done

File: src/dataloader.cpp (hunk 1 above)
Comment (TolyaTalamanov) [reply src/dataloader.cpp]: done

File: tests/unit/test_networks.cpp [hunk 5]
```diff
@@ -5,31 +5,52 @@
...
 
         dw::reference::CPUReLUForward(linear_out0.data(), relu_out1.data(), relu_out1.total());
 
-        dw::reference::CPULinearForward(relu_out1.data(), W1.data(), linear_out2.data(),
+        dw::reference::CPULinearForward(relu_out1.data(), expected_W1.data(), linear_out2.data(),
                                         batch_size, mid_features, out_features);
-        dw::reference::CPULinearAddBias(b1.data(), linear_out2.data(), batch_size, out_features);
+        dw::reference::CPULinearAddBias(expected_b1.data(), linear_out2.data(), batch_size, out_features);
 
         deepworks::reference::CPUSoftmaxForward(linear_out2.data(), output.data(),
```
Comment (RodionovF) [inline tests/unit/test_networks.cpp:55]: dw::

Comment (TolyaTalamanov) [reply tests/unit/test_networks.cpp]: Done

**Comment by TolyaTalamanov** [comment]
@apertovs Let's merge it ?

**Review by apertovs — APPROVED** [review]

**Pull request merged**